### DISCUSSION TITLE
Error	CS1061	'ISetupActivity<If>' does not contain a definition for 'When'

### DISCUSSION THREAD

**Discussion by chavalixebia** [elsa-workflows/elsa-core#5596] [Q&A]
Getting below error for When() condition in elsa 3.1.2

Error	CS1061	'ISetupActivity<If>' does not contain a definition for 'When' and no accessible extension method 'When' accepting a first argument of type 'ISetupActivity<If>' could be found (are you missing a using directive or an assembly reference?)

public class CustomWorkflow : IWorkflow
{
    public void Build(IWorkflowBuilder builder)
    {
        builder
            .StartWith<Activity1>()
            .Then<Activity2>()
            // Add a conditional activity here
            .Add<If>(activity => activity
                .WithCondition(context => context.GetInput<bool>("Flag"))
                .When(OutcomeNames.True).Then<Activity3>() // This will execute if the flag is true
                .When(OutcomeNames.False).Then<Activity4>()) // This will execute if the flag is false
            .Then<Activity5>(); // Continue with Activity5 after the conditional branch
    }
}

**Comment by glime-ai** [comment]
Based on the error message you're encountering, it seems that the `When` method is not being recognized for the `If` activity in your Elsa workflow. This could be due to a few reasons, such as missing using directives, incorrect method usage, or API changes between Elsa versions.

In Elsa 3.x, the way to define conditional flows has changed compared to earlier versions. The `When` method is not used in the same way as it was in Elsa 2.x. Instead, you should use the `When` method on the `Outcome` of an activity to branch the workflow based on the outcome.

Here's how you can define a conditional flow using the `If` activity in Elsa 3.x:

```csharp
public class CustomWorkflow : IWorkflow
{
    public void Build(IWorkflowBuilder builder)
    {
        builder
            .StartWith<Activity1>()
            .Then<Activity2>()
            // Add a conditional activity here
            .Add<IfElse>(ifElse => ifElse
                .WithCondition(context => context.GetInput<bool>("Flag"))
                .When(OutcomeNames.True).Then<Activity3>() // This will execute if the flag is true
                .When(OutcomeNames.False).Then<Activity4>()) // This will execute if the flag is false
            .Then<Activity5>(); // Continue with Activity5 after the conditional branch
    }
}
```

Please note that the `If` activity might be named `IfElse` or similar in Elsa 3.x, and you should ensure that you are using the correct activity type and method signatures as per the Elsa 3.x documentation.

If you continue to face issues, please refer to the official Elsa documentation or check the Elsa GitHub repository for examples and guidance on how to use the `If` activity in Elsa 3.x.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).
